kernel/queue.c

```diff
@@ -293,45 +293,32 @@ int k_queue_merge_slist(struct k_queue *queue, sys_slist_t *list)
 }
 
 #if defined(CONFIG_POLL)
-static void *k_queue_poll(struct k_queue *queue, s32_t timeout)
+static void *k_queue_poll(struct k_queue *queue, k_timeout_t timeout)
 {
 	struct k_poll_event event;
-	int err, elapsed = 0, done = 0;
+	int err;
 	k_spinlock_key_t key;
 	void *val;
-	u32_t start;
 
 	k_poll_event_init(&event, K_POLL_TYPE_FIFO_DATA_AVAILABLE,
 			  K_POLL_MODE_NOTIFY_ONLY, queue);
 
-	if (timeout != K_FOREVER) {
-		start = k_uptime_get_32();
-	}
-
-	do {
-		event.state = K_POLL_STATE_NOT_READY;
-
-		err = k_poll(&event, 1, timeout - elapsed);
-
-		if (err && err != -EAGAIN) {
-			return NULL;
-		}
+	event.state = K_POLL_STATE_NOT_READY;
+	err = k_poll(&event, 1, timeout);
```

Comment:
can you explain why loop is no longer needed?

Reply:
It was never needed.  The assumption seemed to be that k_poll() might fail spuriously and be retried, but k_poll() doesn't fail in a way that requires retries.  We should be propagating any internal errors, not hiding them.

And in this case fixing the design bug was easier than trying to refactor it to do retries over a timeout.